<a href="https://colab.research.google.com/github/Al-ameen007/CNN/blob/main/CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Importing the data and some preprocessing

In [1]:
import numpy as np
from matplotlib import pyplot as plt
import tensorflow
from tensorflow import keras

In [2]:
from tensorflow.keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [3]:
# Normalizing the data and one-hot encoding the labels
from tensorflow.keras.utils import to_categorical
X_train = X_train.astype('float32') / 255.0
X_test = X_test.astype('float32') / 255.0
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [4]:
# Splitting the data into training and validation sets as we want to save the test set for the evaluation of the best model
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.15, random_state=42)

In [5]:
from keras import models
from keras import layers
from tensorflow.keras import optimizers

In [6]:
saved_models = []
accuracies = []

#Base Model

In [7]:
model = models.Sequential()
model.add(layers.Conv2D(64, (5, 5), strides=(2, 2),activation='relu', kernel_initializer='truncated_normal', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu', kernel_initializer='he_uniform'))
model.add(layers.Dense(10, activation='softmax'))

In [8]:
opt = optimizers.SGD(learning_rate=0.0001, momentum=0.9)
opt_1 = optimizers.SGD(learning_rate=0.05, momentum=0.9)
opt_2 = optimizers.SGD(learning_rate=0.001, momentum=0.9)
opt_3 = optimizers.SGD(learning_rate=0.01, momentum=0.9)
opt_4 = optimizers.Adam(learning_rate=0.01, beta_1=0.9, beta_2=0.999)
opt_5 = optimizers.RMSprop(learning_rate=0.01, rho=0.9, momentum=0.1,)

In [37]:
def evaluate_model(saved_models, X, y):
  best_model, best_model_idx = saved_models[0], 0
  best_accuracy= (best_model.evaluate(X, y, verbose=1))[1]
  for i in range(1, len(saved_models)):
    loss, acc = saved_models[i].evaluate(X, y, verbose=1)
    if acc > best_accuracy:
      best_model, best_model_idx , best_accuracy= saved_models[i], i, acc
  print('model: ' + str(best_model_idx + 1), 'accuracy: ' + str('%.2f'%(best_accuracy * 100) + ' %'))
  return best_model, acc

In [10]:
def summaries(saved_models):
  for model in saved_models:
    model.summary()

#Trying differnt number of epochs

In [11]:
model_1 = keras.models.clone_model(model)
model_1.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_1.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_1)

Epoch 1/10
1594/1594 [==============================] - 20s 12ms/step - loss: 2.0366 - accuracy: 0.4922 - val_loss: 1.6412 - val_accuracy: 0.6759
Epoch 2/10
1594/1594 [==============================] - 14s 9ms/step - loss: 1.1477 - accuracy: 0.7566 - val_loss: 0.7950 - val_accuracy: 0.8073
Epoch 3/10
1594/1594 [==============================] - 14s 9ms/step - loss: 0.6402 - accuracy: 0.8387 - val_loss: 0.5403 - val_accuracy: 0.8599
Epoch 4/10
1594/1594 [==============================] - 13s 8ms/step - loss: 0.4833 - accuracy: 0.8723 - val_loss: 0.4399 - val_accuracy: 0.8812
Epoch 5/10
1594/1594 [==============================] - 14s 9ms/step - loss: 0.4127 - accuracy: 0.8871 - val_loss: 0.3872 - val_accuracy: 0.8922
Epoch 6/10
1594/1594 [==============================] - 13s 8ms/step - loss: 0.3714 - accuracy: 0.8967 - val_loss: 0.3513 - val_accuracy: 0.9024
Epoch 7/10
1594/1594 [==============================] - 13s 8ms/step - loss: 0.3436 - accuracy: 0.9023 - val_loss: 0.3270 - val_a

In [12]:
model_2 = keras.models.clone_model(model)
model_2.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_2.fit(X_train, y_train, epochs=15, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_2)

Epoch 1/15
1594/1594 [==============================] - 14s 9ms/step - loss: 2.0773 - accuracy: 0.3959 - val_loss: 1.6823 - val_accuracy: 0.6861
Epoch 2/15
1594/1594 [==============================] - 13s 8ms/step - loss: 1.1483 - accuracy: 0.7729 - val_loss: 0.7674 - val_accuracy: 0.8137
Epoch 3/15
1594/1594 [==============================] - 13s 8ms/step - loss: 0.6175 - accuracy: 0.8436 - val_loss: 0.5194 - val_accuracy: 0.8616
Epoch 4/15
1594/1594 [==============================] - 13s 8ms/step - loss: 0.4685 - accuracy: 0.8723 - val_loss: 0.4268 - val_accuracy: 0.8834
Epoch 5/15
1594/1594 [==============================] - 14s 9ms/step - loss: 0.4027 - accuracy: 0.8880 - val_loss: 0.3769 - val_accuracy: 0.8974
Epoch 6/15
1594/1594 [==============================] - 13s 8ms/step - loss: 0.3643 - accuracy: 0.8966 - val_loss: 0.3452 - val_accuracy: 0.9037
Epoch 7/15
1594/1594 [==============================] - 14s 9ms/step - loss: 0.3385 - accuracy: 0.9030 - val_loss: 0.3217 - val_ac

In [13]:
model_3 = keras.models.clone_model(model)
model_3.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_3.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_3)

Epoch 1/20
1594/1594 [==============================] - 14s 8ms/step - loss: 2.1850 - accuracy: 0.3206 - val_loss: 1.9642 - val_accuracy: 0.5496
Epoch 2/20
1594/1594 [==============================] - 13s 8ms/step - loss: 1.4464 - accuracy: 0.7207 - val_loss: 0.9474 - val_accuracy: 0.7947
Epoch 3/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.7138 - accuracy: 0.8301 - val_loss: 0.5720 - val_accuracy: 0.8520
Epoch 4/20
1594/1594 [==============================] - 14s 9ms/step - loss: 0.5036 - accuracy: 0.8665 - val_loss: 0.4529 - val_accuracy: 0.8750
Epoch 5/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.4225 - accuracy: 0.8846 - val_loss: 0.3943 - val_accuracy: 0.8904
Epoch 6/20
1594/1594 [==============================] - 14s 9ms/step - loss: 0.3781 - accuracy: 0.8937 - val_loss: 0.3585 - val_accuracy: 0.9003
Epoch 7/20
1594/1594 [==============================] - 14s 9ms/step - loss: 0.3490 - accuracy: 0.9006 - val_loss: 0.3341 - val_ac

In [14]:
best_model, best_accuracy = evaluate_model(saved_models, X_test, y_test)

313/313 [==============================] - 1s 3ms/step - loss: 0.1862 - accuracy: 0.9476
model: 3 accuracy: 94.76 %


In [15]:
summaries(saved_models)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 12, 12, 64)        1664      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 6, 6, 64)         0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 2304)              0         
                                                                 
 dense (Dense)               (None, 64)                147520    
                                                                 
 dense_1 (Dense)             (None, 10)                650       
                                                                 
Total params: 149,834
Trainable params: 149,834
Non-trainable params: 0
__________________________________________________

#Trying different learning rates

Model 3 was our best model so far, using 20 epochs, so we take it and apply different learning rates to see if we can get better results.

In [16]:
model_4 = keras.models.clone_model(best_model)
opt_1 = optimizers.SGD(learning_rate=0.05, momentum=0.9)
model_4.compile(optimizer=opt_1, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_4.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_4)

Epoch 1/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.1658 - accuracy: 0.9486 - val_loss: 0.0748 - val_accuracy: 0.9759
Epoch 2/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0634 - accuracy: 0.9802 - val_loss: 0.0999 - val_accuracy: 0.9727
Epoch 3/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0476 - accuracy: 0.9852 - val_loss: 0.0667 - val_accuracy: 0.9789
Epoch 4/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0352 - accuracy: 0.9890 - val_loss: 0.0562 - val_accuracy: 0.9852
Epoch 5/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0293 - accuracy: 0.9905 - val_loss: 0.0540 - val_accuracy: 0.9856
Epoch 6/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0228 - accuracy: 0.9926 - val_loss: 0.0696 - val_accuracy: 0.9839
Epoch 7/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0167 - accuracy: 0.9946 - val_loss: 0.0629 - val_ac

In [17]:
model_5 = keras.models.clone_model(best_model)
model_5.compile(optimizer=opt_2, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_5.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_5)

Epoch 1/20
1594/1594 [==============================] - 14s 8ms/step - loss: 0.7059 - accuracy: 0.8079 - val_loss: 0.2941 - val_accuracy: 0.9141
Epoch 2/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.2603 - accuracy: 0.9233 - val_loss: 0.2177 - val_accuracy: 0.9383
Epoch 3/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.1897 - accuracy: 0.9447 - val_loss: 0.1551 - val_accuracy: 0.9551
Epoch 4/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.1459 - accuracy: 0.9577 - val_loss: 0.1273 - val_accuracy: 0.9638
Epoch 5/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.1197 - accuracy: 0.9651 - val_loss: 0.1077 - val_accuracy: 0.9691
Epoch 6/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.1024 - accuracy: 0.9705 - val_loss: 0.0990 - val_accuracy: 0.9707
Epoch 7/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0913 - accuracy: 0.9733 - val_loss: 0.0905 - val_ac

In [18]:
model_6 = keras.models.clone_model(best_model)
model_6.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_6.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_6)

Epoch 1/20
1594/1594 [==============================] - 14s 9ms/step - loss: 0.2466 - accuracy: 0.9250 - val_loss: 0.0920 - val_accuracy: 0.9708
Epoch 2/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0718 - accuracy: 0.9785 - val_loss: 0.0598 - val_accuracy: 0.9824
Epoch 3/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0514 - accuracy: 0.9838 - val_loss: 0.0460 - val_accuracy: 0.9868
Epoch 4/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0405 - accuracy: 0.9875 - val_loss: 0.0466 - val_accuracy: 0.9864
Epoch 5/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0317 - accuracy: 0.9906 - val_loss: 0.0412 - val_accuracy: 0.9874
Epoch 6/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0267 - accuracy: 0.9918 - val_loss: 0.0388 - val_accuracy: 0.9891
Epoch 7/20
1594/1594 [==============================] - 13s 8ms/step - loss: 0.0204 - accuracy: 0.9937 - val_loss: 0.0400 - val_ac

In [19]:
best_model, best_accuracy = evaluate_model(saved_models, X_test, y_test)

313/313 [==============================] - 1s 3ms/step - loss: 0.0385 - accuracy: 0.9891
model: 6 accuracy: 98.91 %


In [20]:
summaries(saved_models)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 12, 12, 64)        1664      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 6, 6, 64)         0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 2304)              0         
                                                                 
 dense (Dense)               (None, 64)                147520    
                                                                 
 dense_1 (Dense)             (None, 10)                650       
                                                                 
Total params: 149,834
Trainable params: 149,834
Non-trainable params: 0
__________________________________________________

#Trying different architectures

Our last model was our best model, with 20 epochs and a learning rate of 0.01, we will try and explore different architectures.

In [21]:
model_7 = models.Sequential()
model_7.add(layers.Conv2D(32, (3, 3), activation='relu', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_7.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_7.add(layers.Conv2D(32, (3, 3), activation='relu', kernel_initializer='he_uniform'))
model_7.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_7.add(layers.Flatten())
model_7.add(layers.Dense(64, activation='relu', kernel_initializer='he_uniform'))
model_7.add(layers.Dense(32, activation='relu', kernel_initializer='he_uniform'))
model_7.add(layers.Dense(10, activation='softmax'))

In [22]:
model_7.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_7.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_7)

Epoch 1/20
1594/1594 [==============================] - 35s 21ms/step - loss: 0.1805 - accuracy: 0.9434 - val_loss: 0.0621 - val_accuracy: 0.9809
Epoch 2/20
1594/1594 [==============================] - 34s 22ms/step - loss: 0.0592 - accuracy: 0.9818 - val_loss: 0.0594 - val_accuracy: 0.9829
Epoch 3/20
1594/1594 [==============================] - 36s 22ms/step - loss: 0.0423 - accuracy: 0.9863 - val_loss: 0.0421 - val_accuracy: 0.9877
Epoch 4/20
1594/1594 [==============================] - 35s 22ms/step - loss: 0.0332 - accuracy: 0.9894 - val_loss: 0.0491 - val_accuracy: 0.9858
Epoch 5/20
1594/1594 [==============================] - 35s 22ms/step - loss: 0.0262 - accuracy: 0.9917 - val_loss: 0.0643 - val_accuracy: 0.9812
Epoch 6/20
1594/1594 [==============================] - 35s 22ms/step - loss: 0.0206 - accuracy: 0.9933 - val_loss: 0.0454 - val_accuracy: 0.9881
Epoch 7/20
1594/1594 [==============================] - 35s 22ms/step - loss: 0.0174 - accuracy: 0.9947 - val_loss: 0.0415 -

In [23]:
model_8 = models.Sequential()
model_8.add(layers.Conv2D(16, (3, 3), strides = (2, 2),activation='relu', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_8.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_8.add(layers.Conv2D(32, (3, 3), strides = (2, 2),activation='relu', kernel_initializer='he_uniform'))
model_8.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_8.add(layers.Flatten())
model_8.add(layers.Dense(64, activation='relu', kernel_initializer='he_uniform'))
model_8.add(layers.Dense(32, activation='relu', kernel_initializer='he_uniform'))
model_8.add(layers.Dense(10, activation='softmax'))

In [24]:
model_8.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_8.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_8)

Epoch 1/20
1594/1594 [==============================] - 8s 5ms/step - loss: 0.4609 - accuracy: 0.8489 - val_loss: 0.2070 - val_accuracy: 0.9363
Epoch 2/20
1594/1594 [==============================] - 8s 5ms/step - loss: 0.1761 - accuracy: 0.9468 - val_loss: 0.1362 - val_accuracy: 0.9594
Epoch 3/20
1594/1594 [==============================] - 8s 5ms/step - loss: 0.1358 - accuracy: 0.9589 - val_loss: 0.1484 - val_accuracy: 0.9524
Epoch 4/20
1594/1594 [==============================] - 8s 5ms/step - loss: 0.1218 - accuracy: 0.9624 - val_loss: 0.1166 - val_accuracy: 0.9636
Epoch 5/20
1594/1594 [==============================] - 8s 5ms/step - loss: 0.1063 - accuracy: 0.9671 - val_loss: 0.1374 - val_accuracy: 0.9586
Epoch 6/20
1594/1594 [==============================] - 8s 5ms/step - loss: 0.0995 - accuracy: 0.9686 - val_loss: 0.1316 - val_accuracy: 0.9609
Epoch 7/20
1594/1594 [==============================] - 9s 5ms/step - loss: 0.0907 - accuracy: 0.9718 - val_loss: 0.1318 - val_accuracy:

In [25]:
model_9 = models.Sequential()
model_9.add(layers.Conv2D(32, (3, 3), activation='relu', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_9.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_9.add(layers.Conv2D(32, (5, 5), activation='relu', kernel_initializer='he_uniform'))
model_9.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_9.add(layers.Flatten())
model_9.add(layers.Dense(32, activation='relu', kernel_initializer='he_uniform'))
model_9.add(layers.Dense(10, activation='softmax'))

In [26]:
model_9.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_9.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_9)

Epoch 1/20
1594/1594 [==============================] - 45s 28ms/step - loss: 0.1753 - accuracy: 0.9451 - val_loss: 0.0895 - val_accuracy: 0.9714
Epoch 2/20
1594/1594 [==============================] - 51s 32ms/step - loss: 0.0638 - accuracy: 0.9802 - val_loss: 0.0550 - val_accuracy: 0.9858
Epoch 3/20
1594/1594 [==============================] - 45s 28ms/step - loss: 0.0443 - accuracy: 0.9867 - val_loss: 0.0461 - val_accuracy: 0.9871
Epoch 4/20
1594/1594 [==============================] - 45s 28ms/step - loss: 0.0356 - accuracy: 0.9885 - val_loss: 0.0413 - val_accuracy: 0.9890
Epoch 5/20
1594/1594 [==============================] - 45s 28ms/step - loss: 0.0285 - accuracy: 0.9910 - val_loss: 0.0545 - val_accuracy: 0.9850
Epoch 6/20
1594/1594 [==============================] - 45s 28ms/step - loss: 0.0234 - accuracy: 0.9928 - val_loss: 0.0488 - val_accuracy: 0.9863
Epoch 7/20
1594/1594 [==============================] - 45s 28ms/step - loss: 0.0195 - accuracy: 0.9935 - val_loss: 0.0508 -

In [27]:
model_10 = models.Sequential()
model_10.add(layers.Conv2D(128, (3, 3), activation='relu', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_10.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_10.add(layers.Flatten())
model_10.add(layers.Dense(32, activation='relu', kernel_initializer='he_uniform'))
model_10.add(layers.Dense(10, activation='softmax'))

In [28]:
model_10.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_10.fit(X_train, y_train, epochs=20, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_10)

Epoch 1/20
1594/1594 [==============================] - 57s 36ms/step - loss: 0.2544 - accuracy: 0.9241 - val_loss: 0.1064 - val_accuracy: 0.9677
Epoch 2/20
1594/1594 [==============================] - 56s 35ms/step - loss: 0.0859 - accuracy: 0.9742 - val_loss: 0.0722 - val_accuracy: 0.9779
Epoch 3/20
1594/1594 [==============================] - 55s 35ms/step - loss: 0.0572 - accuracy: 0.9825 - val_loss: 0.0715 - val_accuracy: 0.9768
Epoch 4/20
1594/1594 [==============================] - 55s 35ms/step - loss: 0.0414 - accuracy: 0.9874 - val_loss: 0.0562 - val_accuracy: 0.9827
Epoch 5/20
1594/1594 [==============================] - 55s 35ms/step - loss: 0.0295 - accuracy: 0.9906 - val_loss: 0.0657 - val_accuracy: 0.9809
Epoch 6/20
1594/1594 [==============================] - 55s 35ms/step - loss: 0.0229 - accuracy: 0.9930 - val_loss: 0.0678 - val_accuracy: 0.9802
Epoch 7/20
1594/1594 [==============================] - 55s 35ms/step - loss: 0.0167 - accuracy: 0.9949 - val_loss: 0.0557 -

In [32]:
best_model, best_accuracy = evaluate_model(saved_models, X_test, y_test)
summaries(saved_models)

313/313 [==============================] - 3s 8ms/step - loss: 0.0634 - accuracy: 0.9849
model: 9 accuracy: 99.30 %


In [30]:
summaries(saved_models)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 12, 12, 64)        1664      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 6, 6, 64)         0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 2304)              0         
                                                                 
 dense (Dense)               (None, 64)                147520    
                                                                 
 dense_1 (Dense)             (None, 10)                650       
                                                                 
Total params: 149,834
Trainable params: 149,834
Non-trainable params: 0
__________________________________________________

#Trying different batch sizes

Our last model always seem to be our best model, reaching an accuarcy of 99.3,
we used 20 epochs, a learning rate of 0.01 and our architecture was a single convultion layer containing 128 filters of size (3, 3) and a max pooling layer followed by a densle layer of 32 neurons.
we will countinue with it and decide to change the batch size this time around.

In [33]:
model_11 = keras.models.clone_model(best_model)
model_11.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_11.fit(X_train, y_train, epochs=20, batch_size=16, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_11)

Epoch 1/20
3188/3188 [==============================] - 55s 17ms/step - loss: 0.1383 - accuracy: 0.9569 - val_loss: 0.0786 - val_accuracy: 0.9759
Epoch 2/20
3188/3188 [==============================] - 51s 16ms/step - loss: 0.0491 - accuracy: 0.9849 - val_loss: 0.0419 - val_accuracy: 0.9876
Epoch 3/20
3188/3188 [==============================] - 51s 16ms/step - loss: 0.0342 - accuracy: 0.9892 - val_loss: 0.0430 - val_accuracy: 0.9881
Epoch 4/20
3188/3188 [==============================] - 51s 16ms/step - loss: 0.0272 - accuracy: 0.9911 - val_loss: 0.0451 - val_accuracy: 0.9871
Epoch 5/20
3188/3188 [==============================] - 50s 16ms/step - loss: 0.0181 - accuracy: 0.9946 - val_loss: 0.0462 - val_accuracy: 0.9873
Epoch 6/20
3188/3188 [==============================] - 51s 16ms/step - loss: 0.0151 - accuracy: 0.9950 - val_loss: 0.0480 - val_accuracy: 0.9883
Epoch 7/20
3188/3188 [==============================] - 51s 16ms/step - loss: 0.0141 - accuracy: 0.9955 - val_loss: 0.0473 -

In [34]:
model_12 = keras.models.clone_model(best_model)
model_12.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_12.fit(X_train, y_train, epochs=20, batch_size=64, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_12)

Epoch 1/20
797/797 [==============================] - 41s 50ms/step - loss: 0.1911 - accuracy: 0.9400 - val_loss: 0.0716 - val_accuracy: 0.9778
Epoch 2/20
797/797 [==============================] - 40s 50ms/step - loss: 0.0657 - accuracy: 0.9798 - val_loss: 0.0645 - val_accuracy: 0.9803
Epoch 3/20
797/797 [==============================] - 40s 50ms/step - loss: 0.0470 - accuracy: 0.9855 - val_loss: 0.0580 - val_accuracy: 0.9826
Epoch 4/20
797/797 [==============================] - 40s 50ms/step - loss: 0.0369 - accuracy: 0.9886 - val_loss: 0.0413 - val_accuracy: 0.9880
Epoch 5/20
797/797 [==============================] - 40s 50ms/step - loss: 0.0298 - accuracy: 0.9910 - val_loss: 0.0432 - val_accuracy: 0.9866
Epoch 6/20
797/797 [==============================] - 40s 50ms/step - loss: 0.0254 - accuracy: 0.9921 - val_loss: 0.0458 - val_accuracy: 0.9868
Epoch 7/20
797/797 [==============================] - 40s 50ms/step - loss: 0.0211 - accuracy: 0.9937 - val_loss: 0.0535 - val_accuracy:

In [38]:
best_model, best_accuracy = evaluate_model(saved_models, X_test, y_test)

313/313 [==============================] - 2s 8ms/step - loss: 0.0385 - accuracy: 0.9901
model: 9 accuracy: 99.30 %


In [36]:
summaries(saved_models)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 12, 12, 64)        1664      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 6, 6, 64)         0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 2304)              0         
                                                                 
 dense (Dense)               (None, 64)                147520    
                                                                 
 dense_1 (Dense)             (None, 10)                650       
                                                                 
Total params: 149,834
Trainable params: 149,834
Non-trainable params: 0
__________________________________________________

#Trying different activation functions

It seems taht we lucked out and started with the perfect batch size, because model 9 is still our best model, we will try and experiment with different activation functions to see if we can get a better result.

In [32]:
model_13 = models.Sequential()
model_13.add(layers.Conv2D(32, (3, 3), activation='tanh', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_13.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_13.add(layers.Conv2D(32, (5, 5), activation='tanh', kernel_initializer='he_uniform'))
model_13.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_13.add(layers.Flatten())
model_13.add(layers.Dense(32, activation='tanh', kernel_initializer='he_uniform'))
model_13.add(layers.Dense(10, activation='softmax'))

In [33]:
model_13.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_13.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_13)

Epoch 1/10
1594/1594 [==============================] - 57s 36ms/step - loss: 0.1839 - accuracy: 0.9484 - val_loss: 0.0678 - val_accuracy: 0.9802
Epoch 2/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0630 - accuracy: 0.9818 - val_loss: 0.0536 - val_accuracy: 0.9831
Epoch 3/10
1594/1594 [==============================] - 51s 32ms/step - loss: 0.0440 - accuracy: 0.9869 - val_loss: 0.0436 - val_accuracy: 0.9880
Epoch 4/10
1594/1594 [==============================] - 53s 33ms/step - loss: 0.0326 - accuracy: 0.9905 - val_loss: 0.0440 - val_accuracy: 0.9864
Epoch 5/10
1594/1594 [==============================] - 47s 29ms/step - loss: 0.0242 - accuracy: 0.9933 - val_loss: 0.0510 - val_accuracy: 0.9862
Epoch 6/10
1594/1594 [==============================] - 47s 29ms/step - loss: 0.0189 - accuracy: 0.9950 - val_loss: 0.0428 - val_accuracy: 0.9871
Epoch 7/10
1594/1594 [==============================] - 47s 29ms/step - loss: 0.0141 - accuracy: 0.9967 - val_loss: 0.0423 -

In [35]:
model_14 = models.Sequential()
model_14.add(layers.Conv2D(32, (3, 3), activation='selu', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_14.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_14.add(layers.Conv2D(32, (5, 5), activation='selu', kernel_initializer='he_uniform'))
model_14.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_14.add(layers.Flatten())
model_14.add(layers.Dense(32, activation='selu', kernel_initializer='he_uniform'))
model_14.add(layers.Dense(10, activation='softmax'))

In [36]:
model_14.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_14.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_14)

Epoch 1/10
1594/1594 [==============================] - 51s 32ms/step - loss: 0.1583 - accuracy: 0.9529 - val_loss: 0.0946 - val_accuracy: 0.9739
Epoch 2/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0604 - accuracy: 0.9807 - val_loss: 0.0617 - val_accuracy: 0.9821
Epoch 3/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0428 - accuracy: 0.9871 - val_loss: 0.0558 - val_accuracy: 0.9842
Epoch 4/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0299 - accuracy: 0.9905 - val_loss: 0.0528 - val_accuracy: 0.9854
Epoch 5/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0224 - accuracy: 0.9930 - val_loss: 0.0643 - val_accuracy: 0.9829
Epoch 6/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0165 - accuracy: 0.9941 - val_loss: 0.0497 - val_accuracy: 0.9878
Epoch 7/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.0131 - accuracy: 0.9957 - val_loss: 0.0526 -

In [37]:
from keras.layers.advanced_activations import LeakyReLU
model_15 = models.Sequential()
model_15.add(layers.Conv2D(32, (3, 3), kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model_15.add(LeakyReLU(alpha=0.05))
model_15.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_15.add(layers.Conv2D(32, (5, 5), kernel_initializer='he_uniform'))
model_15.add(LeakyReLU(alpha=0.05))
model_15.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model_15.add(layers.Flatten())
model_15.add(layers.Dense(32, kernel_initializer='he_uniform'))
model_15.add(LeakyReLU(alpha=0.05))
model_15.add(layers.Dense(10, activation='softmax'))

In [38]:
model_15.compile(optimizer=opt_3, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_15.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_15)

Epoch 1/10
1594/1594 [==============================] - 49s 30ms/step - loss: 0.1952 - accuracy: 0.9378 - val_loss: 0.0816 - val_accuracy: 0.9762
Epoch 2/10
1594/1594 [==============================] - 48s 30ms/step - loss: 0.0657 - accuracy: 0.9796 - val_loss: 0.0531 - val_accuracy: 0.9832
Epoch 3/10
1594/1594 [==============================] - 48s 30ms/step - loss: 0.0458 - accuracy: 0.9859 - val_loss: 0.0429 - val_accuracy: 0.9873
Epoch 4/10
1594/1594 [==============================] - 48s 30ms/step - loss: 0.0352 - accuracy: 0.9890 - val_loss: 0.0580 - val_accuracy: 0.9838
Epoch 5/10
1594/1594 [==============================] - 48s 30ms/step - loss: 0.0288 - accuracy: 0.9908 - val_loss: 0.0424 - val_accuracy: 0.9890
Epoch 6/10
1594/1594 [==============================] - 48s 30ms/step - loss: 0.0231 - accuracy: 0.9927 - val_loss: 0.0430 - val_accuracy: 0.9871
Epoch 7/10
1594/1594 [==============================] - 48s 30ms/step - loss: 0.0197 - accuracy: 0.9933 - val_loss: 0.0420 -

In [ ]:
best_model, best_accuracy = evaluate_model(saved_models, X_test, y_test)

In [ ]:
summaries(saved_models)

#Trying different optimizers

In [40]:
model_16 = keras.models.clone_model(best_model)
model_16.compile(optimizer=opt_4, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_16.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_16)

Epoch 1/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.1698 - accuracy: 0.9491 - val_loss: 0.1191 - val_accuracy: 0.9642
Epoch 2/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.0935 - accuracy: 0.9726 - val_loss: 0.1090 - val_accuracy: 0.9768
Epoch 3/10
1594/1594 [==============================] - 49s 30ms/step - loss: 0.0887 - accuracy: 0.9764 - val_loss: 0.0927 - val_accuracy: 0.9752
Epoch 4/10
1594/1594 [==============================] - 49s 30ms/step - loss: 0.0824 - accuracy: 0.9781 - val_loss: 0.0939 - val_accuracy: 0.9761
Epoch 5/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.0848 - accuracy: 0.9785 - val_loss: 0.1104 - val_accuracy: 0.9756
Epoch 6/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.0732 - accuracy: 0.9806 - val_loss: 0.0834 - val_accuracy: 0.9814
Epoch 7/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.0826 - accuracy: 0.9797 - val_loss: 0.0992 -

In [41]:
model_17 = keras.models.clone_model(best_model)
model_17.compile(optimizer=opt_5, loss='categorical_crossentropy', metrics=['accuracy'])
history = model_17.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model_17)

Epoch 1/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.1932 - accuracy: 0.9517 - val_loss: 0.1405 - val_accuracy: 0.9726
Epoch 2/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.1487 - accuracy: 0.9693 - val_loss: 0.1114 - val_accuracy: 0.9764
Epoch 3/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.1563 - accuracy: 0.9691 - val_loss: 0.1303 - val_accuracy: 0.9676
Epoch 4/10
1594/1594 [==============================] - 50s 31ms/step - loss: 0.1642 - accuracy: 0.9696 - val_loss: 0.1348 - val_accuracy: 0.9691
Epoch 5/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.1774 - accuracy: 0.9702 - val_loss: 0.3268 - val_accuracy: 0.9729
Epoch 6/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.1881 - accuracy: 0.9686 - val_loss: 0.1251 - val_accuracy: 0.9720
Epoch 7/10
1594/1594 [==============================] - 49s 31ms/step - loss: 0.1937 - accuracy: 0.9663 - val_loss: 0.3031 -

#Trying different dropout rates

In [43]:
from keras.layers import Dropout

In [44]:
model__16 = models.Sequential()
model__16.add(layers.Conv2D(32, (3, 3), kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model__16.add(LeakyReLU(alpha=0.05))
model__16.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model.add(Dropout(0.4))
model__16.add(layers.Conv2D(32, (5, 5), kernel_initializer='he_uniform'))
model__16.add(LeakyReLU(alpha=0.05))
model__16.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model__16.add(layers.Flatten())
model__16.add(layers.Dense(32, kernel_initializer='he_uniform'))
model__16.add(LeakyReLU(alpha=0.05))
model.add(Dropout(0.4))
model__16.add(layers.Dense(10, activation='softmax'))

In [45]:
opt_16 = optimizers.SGD(learning_rate=0.01, momentum=0.9)
model__16.compile(optimizer=opt_16, loss='categorical_crossentropy', metrics=['accuracy'])
history = model__16.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model__16)

Epoch 1/10
1594/1594 [==============================] - 52s 32ms/step - loss: 0.1903 - accuracy: 0.9390 - val_loss: 0.0856 - val_accuracy: 0.9746
Epoch 2/10
1594/1594 [==============================] - 51s 32ms/step - loss: 0.0666 - accuracy: 0.9796 - val_loss: 0.0585 - val_accuracy: 0.9829
Epoch 3/10
1594/1594 [==============================] - 51s 32ms/step - loss: 0.0466 - accuracy: 0.9856 - val_loss: 0.0602 - val_accuracy: 0.9816
Epoch 4/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.0357 - accuracy: 0.9887 - val_loss: 0.0490 - val_accuracy: 0.9864
Epoch 5/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.0290 - accuracy: 0.9907 - val_loss: 0.0474 - val_accuracy: 0.9861
Epoch 6/10
1594/1594 [==============================] - 52s 32ms/step - loss: 0.0239 - accuracy: 0.9925 - val_loss: 0.0410 - val_accuracy: 0.9890
Epoch 7/10
1594/1594 [==============================] - 52s 32ms/step - loss: 0.0200 - accuracy: 0.9931 - val_loss: 0.0515 -

In [46]:
model__17 = models.Sequential()
model__17.add(layers.Conv2D(32, (3, 3), kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model__17.add(LeakyReLU(alpha=0.05))
model__17.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model__17.add(layers.Conv2D(32, (5, 5), kernel_initializer='he_uniform'))
model__17.add(LeakyReLU(alpha=0.05))
model__17.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model__17.add(Dropout(0.75))
model__17.add(layers.Flatten())
model__17.add(layers.Dense(32, kernel_initializer='he_uniform'))
model__17.add(LeakyReLU(alpha=0.05))
model__17.add(Dropout(0.75))
model__17.add(layers.Dense(10, activation='softmax'))

In [47]:
opt_17 = optimizers.SGD(learning_rate=0.01, momentum=0.9)
model__17.compile(optimizer=opt_17, loss='categorical_crossentropy', metrics=['accuracy'])
history = model__17.fit(X_train, y_train, epochs=10, batch_size=32, shuffle=True,validation_data=(X_val, y_val), verbose=1)
saved_models.append(model__17)

Epoch 1/10
1594/1594 [==============================] - 53s 33ms/step - loss: 1.3637 - accuracy: 0.5229 - val_loss: 0.3583 - val_accuracy: 0.9178
Epoch 2/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.8581 - accuracy: 0.7139 - val_loss: 0.2509 - val_accuracy: 0.9318
Epoch 3/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.7381 - accuracy: 0.7610 - val_loss: 0.1655 - val_accuracy: 0.9560
Epoch 4/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.6770 - accuracy: 0.7819 - val_loss: 0.1448 - val_accuracy: 0.9626
Epoch 5/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.6384 - accuracy: 0.7944 - val_loss: 0.1310 - val_accuracy: 0.9646
Epoch 6/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.6101 - accuracy: 0.8014 - val_loss: 0.1232 - val_accuracy: 0.9680
Epoch 7/10
1594/1594 [==============================] - 52s 33ms/step - loss: 0.6031 - accuracy: 0.8057 - val_loss: 0.1369 -

#Models Evaluation

In [48]:
from operator import itemgetter
for i in range(0, len(saved_models)):
  loss, acc = saved_models[i].evaluate(X_test, y_test, verbose = 0)
  accuracies.append(acc)
  print('models: ' + str(i + 1), '%.2f'%(acc * 100) + '%')

models: 1 98.40%
models: 2 97.92%
models: 3 98.93%
models: 4 98.79%
models: 5 96.73%
models: 6 98.99%
models: 7 98.76%
models: 8 98.83%
models: 9 98.67%
models: 10 98.75%
models: 11 98.89%
models: 12 98.57%
models: 13 99.01%
models: 14 98.20%
models: 15 94.99%
models: 16 98.90%
models: 17 95.93%
